Honor 'permissive' setting when creating marc4j reader

adjam opened the pull request:
Current implementation always creates an instance of MarcPermissiveStreamReader, which, regardless of the 'permissive' flag, uses different code to process records.  For those who want the stricter reader, it should be possible to create an ordinary MarcStreamReader.

adjam committed:
Add marc4j_reader.class and permissive setting

adjam commented:
Looking more closely, the original passes in the value of the `marc4j_reader.permissive` setting to the constructor of `MarcPermissiveStreamReader` -- it seems that a better fix would be one that allows one to choose any of the following options:

- A (base) `MarcStreamReader`
- A `MarcPermissiveStreamReader` with `permissive` == `true`
- A `MarcPermissiveStreamReader` with `permissive` == `false`

jrochkind commented:
Sorry, just noticed this!  I was not watching this repo for some reason, maybe I was hoping @billdueber would be. 

If you're still around @adjam, can you rebase to not have that merge commit in there; take out the version bump (we'll handle that separately), and if you wanna be REALLY helpful add a CHANGES.md file?

jrochkind commented:
Oh wait, I'm looking more closely -- the permissive setting wasn't a no-op, as you discovered, it was a permissive flag to `MarcPermissiveStreamReader`.   It looks like you added an ability to set the actual class... and left the permissive flag alone in the end?

I'm confused whats' going on now. This also needs some docs added to the file to explain the meanings of the various configs, including the new one you added!  And maybe rebase into fewer commits?

Commit:
add CHANGES.md

adjam commented:
Part of this is my initial confusion over the meaning of the `permissive` setting.  We had (a while ago) a use case where the `MarcPermissiveStreamReader` class couldn't process records that the `MarcStreamReader` class could -- they're ... wrong records (contain tags with uppercase subfield codes).  So I've rebased and added the CHANGES.md that hopefully explains the new setting and how it relates to the older setting.

As it turns out, the institution with the funky records is probably going to use MRI, but I thought the tweak might be useful for others.

jrochkind commented:
Thanks @adjam!  Am I reading it correctly that this presently has no backwards incompats, all existing settings do exactly what they did before, you just added a new one?  Then seems fine to merge!

adjam commented:
@jrochkind yes, the only functional difference is you can add a setting to choose the `MarcStreamReader` class.

jrochkind commented:
Added mention of the new setting in README, where I noticed just before I was about to release it oughta be. 

Tried to release a 1.1.0, with this change -- but discovered I don't have gem push rights in rubygems for this gem. 

@billdueber, you are the only gem owner in rubygems for this gem. https://rubygems.org/gems/traject-marc4j_reader/]

Bill, you need to add me as an owner:

    gem owner traject-marc4j_reader --add jonathan@dnil.net

Or release this yourself -- what's in the repo master is ready, you just gotta `rake release` I think (not sure if you need to be under jruby when you do it or not). 

Or possibly both of those things. :)

jrochkind commented:
This reminds me that this was ALL Bill before, I've never actually touched this code before, heh.

billdueber commented:
@jrochkind  Finally saw this and did it. jonathan@dnil.net is now an owner.

jrochkind commented:
@adjam traject-marc4j_reader 1.1.0 released, which oughta include this. Thanks!
